### PR TITLE
Add/order model trash untrash

### PR DESCRIPTION
**Pull request by barryhughes**
It is possible to trash orders, regardless of data-store, by calling `$order->delete()`. However, there is not currently a uniform way of recovering orders from the trash:

- For CPT orders, one should use `wp_untrash_post()`
- For HPOS orders, one should use `$order->get_data_store->untrash_order()`

It therefore makes sense to have a single path for untrashing, as an analog of `$order->delete()`. Additionally, this will make it easier to test scenarios [where orders move in and out of the trash](https://github.com/woocommerce/woocommerce/pull/37842) across data-stores. With those things in mind, this PR introduces `$order->untrash()`.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

This change does not impact any user-facing functionality. To compliment code review, though, you may wish to do some testing via WP Shell or similar:

```
>>> $order_id = 12345;
>>> $order = wc_get_order( $order_id )
=> Automattic\WooCommerce\Admin\Overrides\Order {#6136
     +refunds: null,
     +customer_id: null,
   }

>>> $order->get_status()
=> "pending"

>>> $order->delete()
=> true

>>> $order->get_status()
=> "trash"

>>> $order = wc_get_order( $order_id )
=> Automattic\WooCommerce\Admin\Overrides\Order {#6141
     +refunds: null,
     +customer_id: null,
   }

>>> $order->untrash()
=> true

>>> $order->get_status()
=> "pending"
```

☝🏼 As you may observe, it is necessary to obtain a fresh instance of the order object after trashing it, before it can be untrashed. This is because, after calling `$order->delete()`, the order ID is set to `(int) 0` (an existing behavior).

### CODE REVIEW

File: plugins/woocommerce/includes/class-wc-order.php
Comment (vedanshujain) [inline plugins/woocommerce/includes/class-wc-order.php:2296]: perhaps we should do a method_exist check here and fallback to wp_untrash_post for custom implementation of a data store which does not have yet this method. otherwise this will cause fatal. wdyt?

**Review by vedanshujain — COMMENTED** [review]
Looks good to me, left a minor suggestion.

File: plugins/woocommerce/includes/class-wc-order.php
Comment (barryhughes) [reply plugins/woocommerce/includes/class-wc-order.php:2296]: I'm not sure I see the potential for a fatal: when we call a method on the data store it will be routed through [__call()](https://github.com/woocommerce/woocommerce/blob/7.7.2/plugins/woocommerce/includes/class-wc-data-store.php#L220-L226), which takes care of checking if the method exists (if it doesn't exist, it returns null which would then be cast to bool false).

This seems both safe and reasonable, since this change means we will have `untrash_order` in both of the official data-stores (it already exists for the HPOS store) and, for any custom data-stores without it, we report that we were unable to trash the order (by returning false).

You can bear this out by switching to the CPT store, and commenting out this method (or just changing the method name). Then, run this code with an actual order ID:

```php
var_dump( wc_get_order( 12345 )->untrash() );
```

The result should be `(bool) false` without a fatal error.

Comment (vedanshujain) [reply plugins/woocommerce/includes/class-wc-order.php:2296]: Ah, you are right, I forgot that this goes via a `__call` magic method.

**Comment by vedanshujain** [comment]
@barryhughes You have kept this as a draft, are you waiting on something or is it good to merge?

**Comment by barryhughes** [comment]
Sorry, that was not intentional. Should be good to merge once we're happy.

**Review by vedanshujain — APPROVED** [review]
LGTM!

**Pull request merged**
